File: game_plugin_interface/src/sc/protocol/LobbyProtocol.java
package sc.protocol;

import java.util.Arrays;
import java.util.Collection;

import sc.protocol.requests.AuthenticateRequest;
import sc.protocol.requests.CancelRequest;
import sc.protocol.requests.FreeReservationRequest;
import sc.protocol.requests.JoinPreparedRoomRequest;
import sc.protocol.requests.JoinRoomRequest;
import sc.protocol.requests.ObservationRequest;
import sc.protocol.requests.PauseGameRequest;
import sc.protocol.requests.PrepareGameRequest;
import sc.protocol.requests.StepRequest;
import sc.protocol.responses.ErrorResponse;
import sc.protocol.responses.GamePausedEvent;
import sc.protocol.responses.JoinGameResponse;
import sc.protocol.responses.LeftGameEvent;
import sc.protocol.responses.MementoPacket;
import sc.protocol.responses.PrepareGameResponse;
import sc.protocol.responses.RoomPacket;
import sc.shared.GameResult;
import sc.shared.PlayerScore;
import sc.shared.ScoreAggregation;
import sc.shared.ScoreCause;
import sc.shared.ScoreDefinition;
import sc.shared.ScoreFragment;
import sc.shared.SlotDescriptor;

import com.thoughtworks.xstream.XStream;

public abstract class LobbyProtocol
{
	public static XStream registerMessages(XStream xStream)
	{
		registerAdditionalMessages(xStream, Arrays.asList(new Class<?>[] {
				ErrorResponse.class, GamePausedEvent.class,
				JoinGameResponse.class, LeftGameEvent.class,
				MementoPacket.class, PrepareGameResponse.class,
				RoomPacket.class }));

		registerAdditionalMessages(xStream, Arrays.asList(new Class<?>[] {
				AuthenticateRequest.class, CancelRequest.class,
				FreeReservationRequest.class, JoinPreparedRoomRequest.class,
				JoinRoomRequest.class, ObservationRequest.class,
				PauseGameRequest.class, PrepareGameRequest.class,
				StepRequest.class }));
		
		registerAdditionalMessages(xStream, Arrays.asList(new Class<?>[] {
				GameResult.class, PlayerScore.class, ScoreAggregation.class,
				ScoreCause.class, ScoreDefinition.class, ScoreFragment.class,
				SlotDescriptor.class }));

		return xStream;
	}

	public static XStream registerAdditionalMessages(XStream xstream,
			Collection<Class<?>> protocolClasses)
	{
		for (Class<?> clazz : protocolClasses)
		{
			xstream.processAnnotations(clazz);
		}

		return xstream;
	}
}

File: sdk/src/main/server-api/sc/protocol/LobbyProtocol.kt
package sc.protocol

import com.thoughtworks.xstream.XStream
import sc.api.plugins.Team
import sc.protocol.requests.*
import sc.protocol.responses.*
import sc.protocol.room.*
import sc.shared.*

object LobbyProtocol {
    
    @JvmStatic
    fun registerMessages(xStream: XStream): XStream {
        // Requests
        registerAdditionalMessages(xStream, listOf(
                AuthenticateRequest::class.java,
                CancelRequest::class.java,
                JoinGameRequest::class.java,
                JoinPreparedRoomRequest::class.java,
                JoinRoomRequest::class.java,
                ObservationRequest::class.java,
                PauseGameRequest::class.java,
                PlayerScoreRequest::class.java,
                PrepareGameRequest::class.java,
                StepRequest::class.java,
                TestModeRequest::class.java,
        ))
    
        // Responses
        registerAdditionalMessages(xStream, listOf(
                ErrorPacket::class.java,
                GamePreparedResponse::class.java,
                JoinedRoomResponse::class.java,
                ObservationResponse::class.java,
                PlayerScoreResponse::class.java,
                RoomWasJoinedEvent::class.java,
                TestModeResponse::class.java,
        ))
        
        // Messages
        registerAdditionalMessages(xStream, listOf(
                RemovedFromGame::class.java,
                RoomPacket::class.java,
                ErrorMessage::class.java,
                GamePaused::class.java,
                MementoMessage::class.java,
                MoveRequest::class.java,
                WelcomeMessage::class.java,
        ))
        
        // Data
        registerAdditionalMessages(xStream, listOf(
                GameResult::class.java,
                PlayerScore::class.java,
                ScoreAggregation::class.java,
                Team::class.java,
                ScoreCause::class.java,
                ScoreDefinition::class.java,
                ScoreFragment::class.java,
                SlotDescriptor::class.java,
                Score::class.java,
                ScoreValue::class.java,
        ))
        
        return xStream
    }
    
    @JvmStatic
    fun registerAdditionalMessages(xStream: XStream, protocolClasses: Collection<Class<*>>?): XStream {
        if (protocolClasses != null) {
            for (clazz in protocolClasses) {
                xStream.processAnnotations(clazz)
            }
        }
        return xStream
    }
    
}
